Delete old persistence schemas in redshift

dpsutton opened the pull request:
Reminder that recently a change occurred that populates a cache_info table in each persisted schema:

```sql
-- postgres/redshift
test-data=# select * from metabase_cache_424a9_379.cache_info ;
       key        |                value
------------------+--------------------------------------
 settings-version | 1
 created-at       | 2023-03-29T14:16:24.849866Z
 instance-uuid    | 407e4ba8-2bab-470f-aeb5-9fc63fd18c4e
 instance-name    | Metabase Test
(4 rows)
```

So for each cache schema, we can classify it as
- old style (a commit before this change. more and more infrequent)
- new style recent
- new style expired

And we can delete them accordingly in startup:
```
2023-04-11 20:09:03,402 INFO data.redshift :: Dropping expired cache schema: metabase_cache_0149c_359
2023-04-11 20:09:04,733 INFO data.redshift :: Dropping expired cache schema: metabase_cache_0149c_70
2023-04-11 20:09:05,557 INFO data.redshift :: Dropping expired cache schema: metabase_cache_0149c_71
2023-04-11 20:09:06,284 INFO data.redshift :: Dropping expired cache schema: metabase_cache_0149c_90
...
2023-04-11 20:20:33,271 INFO data.redshift :: Dropping expired cache schema: metabase_cache_fe4a7_90
2023-04-11 20:20:34,284 INFO data.redshift :: Dropping old cache schema without `cache_info` table: metabase_cache_8f4b8_358
2023-04-11 20:20:35,076 INFO data.redshift :: Dropping old cache schema without `cache_info` table: metabase_cache_8f4b8_69
...
```

It's possible this will at first cause a few flakes if we are unlucky enough to drop a cache schema without `cache_info` for an instance that is running tests at that point. But the `cache_info` table has been backported so the chances of that become smaller each day.

I've let a week elapse from that change before committing this so hopefully it is not an issue in practice.

Number of queries:
Makes a single query to get all schemas, then for each schema makes a query to classify it. This can be unified into a single query with some shenanigans like

```clojure
(sql/format {:select [:schema :created-at]
             :from   {:union-all
                      (for [schema schemas]
                        {:select [[[:inline schema] :schema]
                                  [{:select [:value]
                                    :from   [(keyword schema "cache_info")]
                                    :where  [:= :key [:inline "created-at"]]}
                                   :created-at]]})}}
            {:dialect :ansi})
```

But i found that this query is extremely slow and does not offer any benefit over the simpler, repeated queries. And as we run this on each commit now, the number of schemas will be far lower and it will be on the order of 5-10 schemas (and therefore queries) and therefore not an issue.

<!-- Reviewable:start -->
- - -
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/metabase/metabase/30013)
<!-- Reviewable:end -->

dpsutton committed:
Delete old persistence schemas in redshift

Reminder that recently a change occurred that populates a cache_info
table in each persisted schema:

```sql
-- postgres/redshift
test-data=# select * from metabase_cache_424a9_379.cache_info ;
       key        |                value
------------------+--------------------------------------
 settings-version | 1
 created-at       | 2023-03-29T14:16:24.849866Z
 instance-uuid    | 407e4ba8-2bab-470f-aeb5-9fc63fd18c4e
 instance-name    | Metabase Test
(4 rows)
```

So for each cache schema, we can classify it as
- old style (a commit before this change. more and more infrequent)
- new style recent
- new style expired

And we can delete them accordingly in startup:
```
2023-04-11 20:09:03,402 INFO data.redshift :: Dropping expired cache schema: metabase_cache_0149c_359
2023-04-11 20:09:04,733 INFO data.redshift :: Dropping expired cache schema: metabase_cache_0149c_70
2023-04-11 20:09:05,557 INFO data.redshift :: Dropping expired cache schema: metabase_cache_0149c_71
2023-04-11 20:09:06,284 INFO data.redshift :: Dropping expired cache schema: metabase_cache_0149c_90
...
2023-04-11 20:20:33,271 INFO data.redshift :: Dropping expired cache schema: metabase_cache_fe4a7_90
2023-04-11 20:20:34,284 INFO data.redshift :: Dropping old cache schema without `cache_info` table: metabase_cache_8f4b8_358
2023-04-11 20:20:35,076 INFO data.redshift :: Dropping old cache schema without `cache_info` table: metabase_cache_8f4b8_69
...
```

It's possible this will at first cause a few flakes if we are unlucky
enough to drop a cache schema without `cache_info` for an instance that
is running tests at that point. But the `cache_info` table has been
backported so the chances of that become smaller each day.

I've let a week elapse from that change before committing this so
hopefully it is not an issue in practice.

Number of queries:
Makes a single query to get all schemas, then for each schema makes a
query to classify it. This can be unified into a single query with some
shenanigans like

```clojure
(sql/format {:select [:schema :created-at]
             :from   {:union-all
                      (for [schema schemas]
                        {:select [[[:inline schema] :schema]
                                  [{:select [:value]
                                    :from   [(keyword schema "cache_info")]
                                    :where  [:= :key [:inline "created-at"]]}
                                   :created-at]]})}}
            {:dialect :ansi})
```

But i found that this query is extremely slow and does not offer any
benefit over the simpler, repeated queries. And as we run this on each
commit now, the number of schemas will be far lower and it will be on
the order of 5-10 schemas (and therefore queries) and therefore not an
issue.

qnkhuat commented on modules/drivers/redshift/test/metabase/test/data/redshift.clj:
```diff
@@ -71,15 +72,87 @@
+   :lacking-created-at should never happen, but if they lack an entry for `created-at`
+   :unknown-error      if an error was thrown while classifying the schema}"
+  [^java.sql.Connection conn schemas]
+  (let [threshold (.minus (java.time.Instant/now)
```
any reason why we can't use `java-time` for these? for example this will be `(t/minus (t/instant) (t/days                           HOURS-BEFORE-EXPIRED-THRESHOLD))`

qnkhuat commented on modules/drivers/redshift/test/metabase/test/data/redshift.clj:
```diff
@@ -71,15 +72,87 @@
+                                                               schema-name))]
+    ;; don't delete unknown-error and recent.
+    (with-open [stmt (.createStatement conn)]
+      (doseq [[collection fmt-str] [[old-convention "Dropping old data schema: %s"]
+                                    [expired "Dropping expired cache schema: %s"]
+                                    [lacking-created-at "Dropping cache without created-at info: %s"]
+                                    [old-style-cache "Dropping old cache schema without `cache_info` table: %s"]]]
+        (doseq [schema collection]
```
```suggestion
(doseq [[collection fmt-str] [[old-convention "Dropping old data schema: %s"]
                                    [expired "Dropping expired cache schema: %s"]
                                    [lacking-created-at "Dropping cache without created-at info: %s"]
                                    [old-style-cache "Dropping old cache schema without `cache_info` table: %s"]]
              schema               collection]
```

this could be nested

qnkhuat commented on modules/drivers/redshift/test/metabase/test/data/redshift.clj:
```diff
@@ -71,15 +72,87 @@
+                                                     (update acc :old conj s)
+                                                     (str/starts-with? s "metabase_cache_")
+                                                     (update acc :cache conj s)
+                                                     :else acc))
```
in which case we'll reach this `else` ?

dpsutton replied:
any other schema in the database (random ones depending on what you are connecting to, system schemas, etc)

qnkhuat commented on modules/drivers/redshift/test/metabase/test/data/redshift.clj:
```diff
@@ -71,15 +72,87 @@
+
+        (group-by classify! schemas)))))
+
 (defn- delete-old-schemas! [^java.sql.Connection conn]
```
IMO it'd be very helpful if you add a for the `delete-old-schemas` method describing what this does and why we do it.

one of the struggles I have reading these test data drivers method is that I don't know why we do a certain thing, and since we don't have tests for it, it's really hard to find the context.

dpsutton committed:
Ngoc's suggestions

- docstring for `delete-old-schemas!`
- combine nested `doseq`
- use java-time over interop with java.time
